workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/selection/VaultWatcherService.kt

```diff
@@ -0,0 +1,148 @@
+            val listOfThings = mutableListOf<StateAndRef<FungibleToken<TokenType>>>()
+            while (existingStates.states.isNotEmpty()) {
+                listOfThings.addAll(existingStates.states as Iterable<StateAndRef<FungibleToken<TokenType>>>)
+                existingStates = appServiceHub.vaultService.queryBy(FungibleToken::class.java, PageSpecification(pageNumber = ++currentPage, pageSize = pageSize))
+            }
+            return VaultObserver(listOfThings, observable)
+        }
+
+    }
+
+
+    //owner -> tokenClass -> tokenIdentifier -> List
+
+
+    init {
+        vaultObserver?.initialValues?.forEach(::addTokenToCache)
+        vaultObserver?.source?.subscribe(::onVaultUpdate)
+    }
+
+    fun onVaultUpdate(t: Vault.Update<FungibleToken<out TokenType>>) {
+        t.consumed.forEach(::removeTokenFromCache)
+        t.produced.forEach(::addTokenToCache)
+    }
+
+    fun removeTokenFromCache(it: StateAndRef<FungibleToken<*>>) {
+        val (owner, type, typeId) = processToken(it.state.data)
+        val tokenSet = getTokenSet(owner, type, typeId)
+        if (tokenSet.remove(it) != true) {
+            LOG.warn("Received a consumption event for a token ${it.ref} which was either not present in the token cache, or was not locked")
+        }
+    }
+
+    fun addTokenToCache(stateAndRef: StateAndRef<FungibleToken<out TokenType>>) {
+        val token = stateAndRef.state.data
+        val (owner, type, typeId) = processToken(token)
+        val tokensForTypeInfo = getTokenSet(owner, type, typeId)
+        val existingMark = tokensForTypeInfo.putIfAbsent(stateAndRef as StateAndRef<FungibleToken<TokenType>>, false)
+        existingMark?.let {
+            LOG.warn("Attempted to overwrite existing token ${stateAndRef.ref} during cache initialization, this suggests incorrect vault behaviours")
+        }
+    }
+
+    fun unlockToken(stateAndRef: StateAndRef<FungibleToken<TokenType>>) {
+        val token = stateAndRef.state.data
+        val (owner, type, typeId) = processToken(token)
+        val tokensForTypeInfo = getTokenSet(owner, type, typeId)
+        tokensForTypeInfo.replace(stateAndRef, true, false)
+    }
+
+    inline fun <T : TokenType> selectTokens(
+            owner: PublicKey,
+            amountRequested: Amount<IssuedTokenType<T>>,
+            predicate: ((StateAndRef<FungibleToken<TokenType>>) -> Boolean) = { true },
+            allowSubSelect: Boolean = false,
+            autoUnlockDelay: Duration = Duration.ofMinutes(5)
+    ): List<StateAndRef<FungibleToken<T>>> {
+        val set = getTokenSet(owner, amountRequested.token.tokenType.tokenClass, amountRequested.token.tokenType.tokenIdentifier)
+        val lockedTokens = mutableListOf<StateAndRef<FungibleToken<TokenType>>>()
+        var amountLocked: Amount<IssuedTokenType<T>> = amountRequested.copy(quantity = 0)
+        for (tokenStateAndRef in set.keys) {
```

Comment:
Not totally sure iterating on a `ConcurrentHashMap` key Set is necessarily a good idea.  What ordering does it have?  No idea how well it performs.  Can you get a concurrent modification exception?

Anyway, all this would be functions on `Bucket` class.